design/design-doc-external-secret-stores.md

```diff
@@ -0,0 +1,655 @@
+API that satisfies the separation of concerns between platform operators and
+consumers which Crossplane already enables today for Kubernetes Secrets. We will
+end up having a unified configuration spec for all external secret store types
+which contains the name field (`name`) and a reference to any additional
+store specific configuration (`configRef`).
+
+This would be enough to uniquely identify and access any secret instance,
+however there could still be some additional attributes or metadata specific to
+store type that might be desired to be set per secret instance. For example,
+`labels`, `annotations` and `type` of the secret in _Kubernetes_; `Tags` and
+`EncryptionKey` in _AWS_. For local Kubernetes, this could be supported by a
+strong typed spec, however, for external stores, to stick with a unified
+interface, we will receive this configuration as arbitrary key value pairs with
+an optional `attributes` field.
+
+**Examples:**
+
+Publish Connection details to a _Kubernetes_ secret (already existing case):
+
+```yaml
+spec:
+  publishConnectionDetailsTo:
+    name: my-db-connection
+    kubernetes:
+      namespace: crossplane-system
+```
+
+Publish Connection details to a _Vault_ secret:
+
+```yaml
+spec:
+  publishConnectionDetailsTo:
+    name: my-db-connection
+    externalStore:
+      configRef:
+        name: vault-default
+        namespace: crossplane-system
+```
+
+Publish Connection details to a _Kubernetes_ secret with some labels:
+
+```yaml
+spec:
+  publishConnectionDetailsTo:
+    name: my-db-connection
+    kubernetes:
+      namespace: crossplane-system
+      labels:
+        environment: production
+      annotations:
+        acme.example.io/secret-type: infrastructure
+```
+
+Publish Connection details to an _AWS Secret Manager_ secret with some tags:
+
+```yaml
+spec:
+  publishConnectionDetailsTo:
+    name: my-db-connection
+    externalStore:
+      configRef:
+        name: aws-secret-manager-platform
+        namespace: crossplane-system
+      attributes:
+        tags:
+          environment: production
+```
+
+Publish Connection details to _Vault_ but **still store it as a Kubernetes
+secret**:
+
+```yaml
+spec:
+  publishConnectionDetailsTo:
+    name: my-db-connection
+    kubernetes:
+      namespace: crossplane-system
+    externalStore:
+      configRef:
+        name: vault-platform
+        namespace: crossplane-system
+```
+
+#### Separation of Concerns: Compositions, Composites and Claims
+
+`StoreConfig` resource will be a _namespaced_ resource to allow platform
+consumers to configure their own secret store of choice. This is aligned with
+the existing approach since Kubernetes secrets are namespaced resources and
+platform operators could specify where should composite resource secrets land
+using the partial input `writeConnectionSecretsToNamespace`. We will similarly
+deprecate this field in favor of `publishConnectionDetailsToStore`.
```

Comment:
I believe it would be important to support multi-tenant scenarios, where, platform operators share a control plane between multiple teams where each team has its own cloud account (e.g. want to use AWS secret store under their account), which could be quite common for big enterprises to manage/track cloud costs. And I remember a very recent discussion with a Platform engineer working with that setup (cc @luebken). The discussion was indeed around `ProviderConfig` and the question was how can I let my platform consumer teams use their own cloud accounts, but it also applies here.

Putting consistency concerns aside, I believe a namespaced `StoreConfig` and a cluster scoped `ClusterStoreConfig` would be ideal. And I am wondering if we can find a _not so great_ solution with some wording trick like:
Cluster scoped: `StoreConfig` and `ProviderConfig`
Namespaced: `TenantStoreConfig` or `NamespacedStoreConfig` and `TenantProviderConfig` or `NamespacedProviderConfig`